teuthology/run.py

```diff
@@ -179,26 +180,89 @@ def validate_tasks(config):
     return config["tasks"]
 
 
+def get_nodes_request(config, machine_type):
+    """
+    Examine each item in a job's 'nodes' stanza. Consolidate those with the
+    same requirements into 'node requests' so that we may later call
+    lock_many() as few times as possible.
+
+    Each resulting request contains role lists, each of which will be mapped to
+    a target when the machines are locked.
+    """
+    request = list()
+    os_specs = OrderedDict()
+    # Group node confs by 'spec'. We use an OrderedDict to contain them
+    # briefly, because it combines the functionality of an ordered set with
+    # key-value storage.
+    for item in config:
+        item['arch'] = item.get('arch')
+        item['machine_type'] = item.get('machine_type', machine_type)
+        spec_key = (
+            item.get('os_type'),
+            item.get('os_version'),
+            item.get('arch'),
+            item.get('machine_type'),
+        )
+        spec_roles = os_specs.get(spec_key, list())
+        assert isinstance(spec_roles, list)
+        spec_roles.append(item['roles'])
+        os_specs[spec_key] = spec_roles
+
+    # Build a 'request' for each 'spec'
+    for spec, roles in os_specs.items():
+        os_type, os_version, arch, machine_type = spec
+        request.append(dict(
+            os_type=os_type or None,
+            os_version=os_version or None,
+            arch=arch or None,
+            machine_type=machine_type,
+            roles=roles,
+        ))
+    return request
+
+
 def get_initial_tasks(lock, config, machine_type):
     init_tasks = [
         {'internal.check_packages': None},
         {'internal.buildpackages_prep': None},
     ]
-    if 'roles' in config and lock:
-        msg = ('You cannot specify targets in a config file when using the ' +
-               '--lock option')
-        assert 'targets' not in config, msg
-        init_tasks.append({'internal.lock_machines': (
-            len(config['roles']), machine_type)})
+
+    target_conflict_msg = 'You cannot specify targets in a config file when' \
+        'using the --lock option'
+    if lock and ('roles' in config or 'nodes' in config):
+        assert 'targets' not in config, target_conflict_msg
+    if lock and 'roles' in config:
+        if 'nodes' in config:
+            log.warn(
```

Comment:
I think we should exit here, It would be a user error to have both and fix them to use only one format